File: python/494.py
"""Project Euler Problem 494: Collatz prefix families.

A Collatz prefix is a Collatz sequence that ends right before a power of 2.
Find the number of distinct families of Collatz prefixes of N terms.

The base count is fibonacci(N). We then find special starting values where
the ordering of terms doesn't follow the standard pattern, and count
additional families from extending those special sequences backward.
"""

from __future__ import annotations
import subprocess
import os


def fibonacci(n):
    """Compute nth Fibonacci number (F_1=1, F_2=1, F_3=2, ...)."""
    a, b = 1, 1
    for _ in range(n - 1):
        a, b = b, a + b
    return a


def is_power_of_2(n):
    """Check if n is a power of 2."""
    return n > 0 and (n & (n - 1)) == 0


def solve():
    """Solve Problem 494."""
    N = 90
    L = 100000
    M = 3 ** 39

    ans = fibonacci(N)

    # Find special sequences.
    # In the Java code, the reference to the collatz list is stored,
    # meaning the FULL list (at the end of the loop) is what gets used.
    # A start is "special" if at any point n > start and r < 1.
    special_seqs = {}
    for start in range(1, L):
        collatz = []
        n = start
        r = 1.0
        is_special = False
        for i in range(N):
            collatz.append(n)
            if n % 2 == 0:
                n //= 2
                r /= 2
            else:
                n = 3 * n + 1
                r *= 3
            if is_power_of_2(n):
                break
            if n > start and r < 1:
                is_special = True
        if is_special:
            special_seqs[start] = collatz  # Store the full list

    # Remove redundant sequences
    non_redundant = {}
    for start, collatz in special_seqs.items():
        is_redundant = False
        for i in range(1, len(collatz)):
            if collatz[i] in special_seqs:
                is_redundant = True
                break
        if not is_redundant:
            non_redundant[start] = collatz

    # Compile C helper if needed
    script_dir = os.path.dirname(os.path.abspath(__file__))
    helper_c = os.path.join(script_dir, "494_helper.c")
    helper_bin = os.path.join(script_dir, "494_helper")
    if not os.path.exists(helper_bin) or os.path.getmtime(helper_c) > os.path.getmtime(helper_bin):
        subprocess.run(["gcc", "-O2", "-o", helper_bin, helper_c], check=True)

    # Count paths for each non-redundant special sequence using C helper
    for start, collatz in non_redundant.items():
        num_steps = N - len(collatz)
        result = subprocess.run(
            [helper_bin, str(start), str(num_steps), "0"],
            capture_output=True, text=True, check=True
        )
        ans += int(result.stdout.strip())

    return ans


def main():
    """Main entry point."""
    result = solve()
    print(result)
    return result


if __name__ == "__main__":
    main()

File: c/494.c
/* Project Euler 494 - Collatz prefix families
 * Translated from python/494.py + python/494_helper.c
 *
 * Self-contained: finds special Collatz sequences and counts paths.
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef unsigned long long ull;

#define NSTEPS 90
#define L 100000

/* Compute M = 3^39 */
static ull M3_39;

void compute_M(void) {
    M3_39 = 1;
    for (int i = 0; i < 39; i++)
        M3_39 *= 3;
}

/* Fibonacci: F_1=1, F_2=1, F_3=2, ... */
ull fibonacci(int n) {
    if (n <= 0) return 0;
    ull a = 1, b = 1;
    for (int i = 2; i < n; i++) {
        ull c = a + b;
        a = b;
        b = c;
    }
    return (n == 1) ? a : b;
}

int is_power_of_2(ull n) {
    return n > 0 && (n & (n - 1)) == 0;
}

ull helper_func(ull n, int numSteps, int nEven) {
    if (numSteps == 0) return 1;
    ull result = helper_func((n * 2) % M3_39, numSteps - 1, 1);
    if (nEven && n % 3 == 1)
        result += helper_func((n - 1) / 3, numSteps - 1, 0);
    return result;
}

/* Store special sequences */
typedef struct {
    ull start;
    ull collatz[NSTEPS + 1];
    int len;
} SeqEntry;

int main() {
    compute_M();

    ull ans = fibonacci(NSTEPS);

    /* Find special sequences */
    SeqEntry *specials = malloc(L * sizeof(SeqEntry));
    int num_specials = 0;

    /* We also need a set of special start values for redundancy check */
    /* Use a simple hash set */
    #define HASH_SIZE 200003
    ull *hash_table = calloc(HASH_SIZE, sizeof(ull));
    char *hash_used = calloc(HASH_SIZE, 1);

    for (ull start = 1; start < L; start++) {
        ull collatz[NSTEPS + 1];
        int len = 0;
        ull n = start;
        double r = 1.0;
        int is_special = 0;

        for (int i = 0; i < NSTEPS; i++) {
            collatz[len++] = n;
            if (n % 2 == 0) {
                n /= 2;
                r /= 2;
            } else {
                n = 3 * n + 1;
                r *= 3;
            }
            if (is_power_of_2(n)) break;
            if (n > start && r < 1.0) is_special = 1;
        }

        if (is_special) {
            specials[num_specials].start = start;
            memcpy(specials[num_specials].collatz, collatz, len * sizeof(ull));
            specials[num_specials].len = len;
            num_specials++;

            /* Add to hash set */
            ull idx = start % HASH_SIZE;
            while (hash_used[idx]) {
                if (hash_table[idx] == start) break;
                idx = (idx + 1) % HASH_SIZE;
            }
            hash_table[idx] = start;
            hash_used[idx] = 1;
        }
    }

    /* Check if a value is in the special set */
    #define in_special(v) ({ \
        int _found = 0; \
        ull _idx = (v) % HASH_SIZE; \
        while (hash_used[_idx]) { \
            if (hash_table[_idx] == (v)) { _found = 1; break; } \
            _idx = (_idx + 1) % HASH_SIZE; \
        } \
        _found; \
    })

    /* Remove redundant sequences */
    for (int si = 0; si < num_specials; si++) {
        int is_redundant = 0;
        for (int i = 1; i < specials[si].len; i++) {
            if (in_special(specials[si].collatz[i])) {
                is_redundant = 1;
                break;
            }
        }
        if (!is_redundant) {
            int num_steps = NSTEPS - specials[si].len;
            ull paths = helper_func(specials[si].start, num_steps, 0);
            ans += paths;
        }
    }

    printf("%llu\n", ans);

    free(specials);
    free(hash_table);
    free(hash_used);
    return 0;
}
